tests/Unit/Framework/Pypp.hpp

```diff
@@ -21,262 +22,311 @@
-                                                            ret, s);
-    }
-    Py_DECREF(func);    // NOLINT
-    Py_DECREF(module);  // NOLINT
-    return return_container;
-  }
-};
-
-template <typename ScalarSpinWeighted>
-struct CallImpl<
-    ScalarSpinWeighted,
-    Requires<(
-        tt::is_a_v<Tensor, ScalarSpinWeighted> or
-        tt::is_std_array_v<ScalarSpinWeighted>) and
-        is_any_spin_weighted_v<typename ScalarSpinWeighted::value_type>>> {
-  template <typename... Args>
-  static ScalarSpinWeighted call(const std::string& module_name,
-                                 const std::string& function_name,
-                                 const Args&... t) {
-    static_assert(sizeof...(Args) > 0,
-                  "Call to python which returns a "
-                  "Scalar<SpinWeighted<ComplexDataVector, N>> "
-                  "should pass at least one argument");
-
-    PyObject* module = PyImport_ImportModule(module_name.c_str());
-    if (module == nullptr) {
-      PyErr_Print();
-      throw std::runtime_error{std::string("Could not find python module.\n") +
-                               module_name};
-    }
-    PyObject* func = PyObject_GetAttrString(module, function_name.c_str());
-    if (func == nullptr or not PyCallable_Check(func)) {
-      if (PyErr_Occurred()) {
-        PyErr_Print();
+    const std::array<size_t, sizeof...(Args)> arg_sizes{
+        {ContainerPackAndUnpack<Args>::get_size(t)...}};
+    const size_t npts = *std::max_element(arg_sizes.begin(), arg_sizes.end());
+    for (size_t i = 0; i < arg_sizes.size(); ++i) {
+      if (arg_sizes[i] != 1 and arg_sizes[i] != npts) {
+        ERROR(
+            "Each argument must return size 1 or "
+            << npts
+            << " (the number of points in the DataVector), but argument number "
+            << i << " has size " << arg_sizes[i]);
       }
-      throw std::runtime_error{"Could not find python function in module.\n"};
     }
-
-    const auto put_container_of_elements_into_container_of_complexdatavector =
-        [](auto& container_vector, const auto& container_complex,
-           const size_t slice_idx) noexcept {
-      ASSERT(slice_idx < container_vector.begin()->size(),
-             "Trying to slice DataVector of size "
-                 << container_vector.begin()->size() << "with slice_idx "
-                 << slice_idx);
-      for (decltype(auto) vector_and_complex_components :
-           boost::combine(container_vector, container_complex)) {
-        boost::get<0>(vector_and_complex_components).data()[slice_idx] =
-            boost::get<1>(vector_and_complex_components);
-      }
-    };
-
-    const size_t npts = get_first_argument(t...).begin()->size();
-    auto return_container = ScalarSpinWeighted{npts};
+    typename ContainerPackAndUnpack<R>::packed_type return_container =
+        ContainerPackAndUnpack<R>::construct_zero_result(npts);
```

Comment:
It's useful since for custom conversion classes there's no other way (that I've thought of) to retrieve the packed type. I'll make sure it's used the way you suggested.